### PR TITLE
#773: Add steps to install Rancher Extensions in an air-gapped environment

### LINKED ISSUE #773: Add steps to install Rancher Extensions in an air-gapped environment
There is work in progress for SURE-6065 to provide a way to install Rancher Extensions in an air-gapped environment. The approach requires an additional step or two on top of the steps provided in [Air-Gapped Helm CLI Install](https://ranchermanager.docs.rancher.com/pages-for-subheaders/air-gapped-helm-cli-install) page. We should capture those steps to install extensions in this page. 

@jordojordo from the UI team should be able to help with the actual steps.

### PR DESCRIPTION
**Pull request by martyav**
Fixes #773, https://github.com/rancher/rancher/issues/42093

> There is work in progress for SURE-6065 to provide a way to install Rancher Extensions in an air-gapped environment. The approach requires an additional step or two on top of the steps provided in [Air-Gapped Helm CLI Install](https://ranchermanager.docs.rancher.com/pages-for-subheaders/air-gapped-helm-cli-install) page. We should capture those steps to install extensions in this page.

Opted instead to list instructions on the Extensions page and link to the relevant section from the air-gapped install guide, as per discussion in the issue thread.

The Extensions page could use a refresh. I asked the Dashboard team about rewording a note that isn't directly related to this issue: https://github.com/rancher/dashboard/issues/9580

### CODE REVIEW

**Review by jordojordo — APPROVED** [review]
lgtm

**Review by btat — APPROVED** [review]

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 1]
```diff
@@ -49,6 +49,30 @@ In v2.7.0, the built-in extensions will not be displayed under the **Available**
 
     ![Reload button](/img/reload-button.png)
 
+### Importing and Installing Extensions in an Air-Gapped Environment
+
+1. Find the address of the container image you want to import as an extension.
```
Comment (rohitsakala) [inline docs/integrations-in-rancher/rancher-extensions.md:54]: Rancher provides some extensions such as Kubewarden and elemental and we need to tell the user about this. The extensions are available in this container image https://hub.docker.com/r/rancher/ui-plugin-catalog/tags

The latest available tag is recommended to be used and imported.

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 2]
```diff
@@ -49,6 +49,30 @@ In v2.7.0, the built-in extensions will not be displayed under the **Available**
...
+
+1. [Create](../how-to-guides/new-user-guides/kubernetes-resources-setup/secrets.md) a registry secret within the `cattle-UI-plugin-system` namespace. Enter the domain of the image address in the **Registry Domain Name** field.
```
Comment (rohitsakala) [inline docs/integrations-in-rancher/rancher-extensions.md:56]: There is not always a secret needed. For example, for publicly available image such as https://hub.docker.com/r/rancher/ui-plugin-catalog/tags, there is no need to secret to be created. Can we mention this somehow ?

**Review by rohitsakala — CHANGES_REQUESTED** [review]
The Uninstall and Upgrade section needs to include how to upgrade an *Extension Catalog* like how to upgrade to the latest tag of the catalog container image

Upgrading can be in two ways right now 
1. Update the tag in the deployment.yaml 
2. Deleting the current *Extension Catalog* and then readding it using the latest tag. 

 and how to uninstall an *Extension Catalog* like deleting the catalog container image. 
 
 Maybe, we can add a new page for airgap?

**Comment by rohitsakala** [comment]
@btat are you going to tackle Uninstalling and upgrade in a separate PR ?

**Comment by martyav** [comment]
@rohitsakala Apologies for the slow response. 

I'm a little confused by the "catalog" terminology. The latest versions of Rancher don't use the term "catalog" in the UI. In public-facing docs we've been using **Apps & Marketplace** instead, starting in v2.5: https://ranchermanager.docs.rancher.com/v2.5/pages-for-subheaders/helm-charts-in-rancher#changes-in-rancher-v25

For reference, this is the latest version of the page that once described the deprecated catalog system: https://ranchermanager.docs.rancher.com/pages-for-subheaders/helm-charts-in-rancher

In the intro of the [Rancher Extensions](https://ranchermanager.docs.rancher.com/integrations-in-rancher/rancher-extensions) page edited here, we distinguish between extensions and apps:

> Extensions are Helm charts that can only be installed once into a cluster; therefore, these charts have been simplified and separated from the general Helm charts listed under Apps.
>
> Examples of built-in Rancher extensions are Fleet, Explorer, and Harvester. Examples of other extensions that use the Extensions API that can be manually added are Kubewarden and Elemental.

So, if we added information about catalogs (Apps), that seems like it should go on https://ranchermanager.docs.rancher.com/pages-for-subheaders/helm-charts-in-rancher in a separate PR. But Kubewarden and Elemental should be described under the Extensions page as they use the Extensions API.

Please correct me if I misunderstood, I want to get everyone on the same page so we can move forward.

**Comment by martyav** [comment]
> I wanted to know if there are sections on how to upgrade `Extensions` ((the helm repository which contains helm charts) to a new version. or how to delete `Extensions` (the helm repository which contains helm charts) ?

We don't seem to. That material should probably go on https://ranchermanager.docs.rancher.com/pages-for-subheaders/helm-charts-in-rancher or a new subpage under that topic.

**Comment by rohitsakala** [comment]
@martyav Why do you think we need to have docs for uninstall and upgrade in a separate section and install in separate section ? They should be in one page or atleast under one section right ?

**Comment by martyav** [comment]
@rohitsakala After reviewing the thread more closely -- the Extensions page in this PR currently describes how to **update** through the UI, but not how to **upgrade**. It also describes how to uninstall extensions, again through the UI. 

If we're adding material that generally applies to deleting/upgrading helm charts or Apps, it should go on https://ranchermanager.docs.rancher.com/pages-for-subheaders/helm-charts-in-rancher. Maybe that's where I misunderstood.

But if it's specific instructions for working with Extensions, it should go on the Extensions page.

**Comment by rohitsakala** [comment]
@martyav So let me also re-analyze this 

**CRUD of Specific Extension** 
- Example -  Kubewarden or Elemental 
- Creation - Documented in https://ranchermanager.docs.rancher.com/v2.8/integrations-in-rancher/rancher-extensions
- Update - Documneted in https://ranchermanager.docs.rancher.com/v2.8/integrations-in-rancher/rancher-extensions
- Deletion - Documented in https://ranchermanager.docs.rancher.com/v2.8/integrations-in-rancher/rancher-extensions



**CRUD of Extensions Repository**

#### Helm Github Repository (Works in non airgap mode only)
- Creation - Already documented in https://ranchermanager.docs.rancher.com/v2.8/integrations-in-rancher/rancher-extensions
- Deletion - Not Documented 
- Update - Not Documented

#### Helm Container Image  (Works in Airgap mode and non airgap mode)
- Creation - Already documented in this PR
- Deletion - Not Documented 
- Updation - Not Documented 


Can you please let me know if this is right?

**Comment by btat** [comment]
Hi @rohitsakala, Marty and I had a chat and your reanalysis sounds right. I've also added additional comments and a question below.

> **CRUD of Specific Extension**

We've been requested to add separate pages for integrations such as Kubewarden and Elemental. For such integrations that use the extension mechanism, the CRUD content will be a subpage of the integration page (I.e., https://ranchermanager.docs.rancher.com/v2.8/integrations-in-rancher/kubewarden will have a subpages added). I think we can keep this part out of this PR's scope.
 
> **CRUD of Extensions Repository**
> #### Helm Github Repository (Works in non airgap mode only)
>
> #### Helm Container Image (Works in Airgap mode and non airgap mode)

I think we should include these 2 areas as part of this PR.

> The Uninstall and Upgrade section needs to include how to upgrade an Extensions like how to upgrade to the latest tag of the catalog container image
>
> Upgrading can be in two ways right now
> 
> 1. Update the tag in the deployment.yaml
> 2. Deleting the current Extensions and then readding it using the latest tag.

Are these 2 methods required for air-gap environments because the Update method listed on Step 5 of [this section](https://ranchermanager.docs.rancher.com/v2.8/integrations-in-rancher/rancher-extensions#installing-extensions) doesn't work for air-gap or are these alternatives/additional methods?

**Comment by rohitsakala** [comment]
@btat Thank you for the reply. 

The point 5 of https://ranchermanager.docs.rancher.com/v2.8/integrations-in-rancher/rancher-extensions#installing-extensions points to the update of a particular extension example kuberwarden. and this is out of the scope of this PR as you mentioned. 

Let me add some information.

**CRUD of Extensions Repository**

#### Helm Github Repository (Works in non airgap mode only)
- Creation - Already documented in https://ranchermanager.docs.rancher.com/v2.8/integrations-in-rancher/rancher-extensions
- Deletion - Not Documented 
- Update - Not Documented

#### Helm Container Image  (Works in Airgap mode and non airgap mode)
- Creation - Already documented in this PR
- Deletion - Not Documented 
- Updation - Right now, only updating from deployment.yaml is possible. In an airgap scenario, the users will have to mirror the image into their private registry before doing the update.


Can you please let me know if this is right?

**Comment by martyav** [comment]
One last question, how would the user "re-add [the extension] using the latest tag", assuming that the user can do so through the UI?

**Comment by rohitsakala** [comment]
@martyav that tag is always associated with the extension repository, not an extension. 

Helm Container Image (Works in Airgap mode and non airgap mode)

Example of an extension repository - `rancher/ui-plugin-catalog:1.0.1` where 1.0.1 is the tag.

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 3]
```diff
@@ -26,7 +26,7 @@ Examples of built-in Rancher extensions are Fleet, Explorer, and Harvester. Exam
 
 :::info
 
-In v2.7.0, the built-in extensions will not be displayed under the **Available** tab. Therefore, you will need to manually add the desired repos to install extensions. We will update the community once these extensions have been pulled out to be available for selection.
+If you've upgraded from an older Rancher version to v2.7.0, uninstalled built-in extensions are no longer listed under the **Available** tab, unless you've manually imported them.
```
Comment (btat) [inline docs/integrations-in-rancher/rancher-extensions.md:29]:
```suggestion
In v2.7.0, the built-in extensions will not be displayed under the **Available** tab. Therefore, you will need to manually add the desired repos to install extensions.
```

The original text is to indicate a limitation found in certain versions (2.7.0-2.7.?) where users would expect something that's built-in to be available for selection. The update here changes the meaning to indicate that there's a change in existing behavior. Also, extensions were introduced in 2.7 so there is no upgrade scenario involved.

**Review by btat — DISMISSED** [review]

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 4]
```diff
@@ -41,17 +41,41 @@ In v2.7.0, the built-in extensions will not be displayed under the **Available**
...
-5. Under the **Available** tab, click **Install** on the desired extension and version as in the example below. Note that you can easily update your extension as the button to **Update** will appear on the extension if one is available.
+5. Under the **Available** tab, click **Install** on the desired extension and version as in the example below. You can also update your extension from this screen, as the button to **Update** will appear on the extension if one is available.
 
     ![Install Kubewarden](/img/install-kubewarden.png)
 
 6. Click the **Reload** page button that will appear after your extension successfully installs. Note that a logged-in user who has just installed an extension will not see a change to the UI **unless** they reload the page.
 
     ![Reload button](/img/reload-button.png)
 
+### Importing and Installing Extensions in an Air-Gapped Environment
```
Comment (rohitsakala) [inline docs/integrations-in-rancher/rancher-extensions.md:52]: Nit: Importing and Installing Extensions Repository in an Air-Gapped Environment

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 5]
```diff
@@ -41,17 +41,41 @@ In v2.7.0, the built-in extensions will not be displayed under the **Available**
...
+
+1. Click **Load**. The extension will now be **Pending**.
+
+1. Return to the **Extensions** page.
+
+1. Select the **Available** tab, and click the **Reload** button to make sure that the list of extensions is up to date.
+
+1. Find the extension you just added, and click the **Install** button.
+
 ## Uninstalling Extensions
```
Comment (rohitsakala) [inline docs/integrations-in-rancher/rancher-extensions.md:76]: Nit: Uninstalling Extensions Repositories

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 6]
```diff
@@ -67,17 +91,28 @@ You must reload the page after disabling extensions or display issues may occur.
...
-Under the **Installed** tab, click the **Rollback** button on the extension you wish to roll back.
+1. Click **☰ > Extensions** under **Configuration**.
+1. Select the **Updates** tab. 
+1. Click **Update**.
 
-![Roll back extensions](/img/roll-back-extension.png)
+If there is a new version of the extension, there will also be an **Update** button visible on the associated card for the extension in the **Available** tab.
 
-:::caution
+### Updating and Upgrading Extensions in an Air-gapped Environment
```
Comment (rohitsakala) [inline docs/integrations-in-rancher/rancher-extensions.md:102]: Updating and Upgrading Extensions Repository

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 7]
```diff
@@ -67,17 +91,28 @@ You must reload the page after disabling extensions or display issues may occur.
...
 
-You must reload the page after rolling back extensions or display issues may occur.
+1. Click **☰**, then click on the name of your local cluster.
+1. From the sidebar, select **Workloads > Deployments**.
+1. From the namespaces dropdown menu, select **All Namespaces**.
+1. Find the **cattle-ui-plugin-system** namespace. 
 
-:::
+<!-- TK Waiting for answer -->
```
Comment (rohitsakala) [inline docs/integrations-in-rancher/rancher-extensions.md:109]: Find the deployment 
ui-plugin-catalog

Click on edit config 
change rancher/ui-plugin-catalog:1.0.1 to rancher/ui-plugin-catalog:1.0.2 (1.0.1 and 1.0.2 are just examples)

click on save.

Comment (rohitsakala) [reply docs/integrations-in-rancher/rancher-extensions.md:109]: Also, they need to mirror the image rancher/ui-plugin-catalog:1.0.2 to their private registry before dpoing this.

**Comment by rohitsakala** [comment]
**CRUD of Extensions Repository**

#### Helm Github Repository (Works in non-airgap mode only)
- Creation - Already documented in https://ranchermanager.docs.rancher.com/v2.8/integrations-in-rancher/rancher-extensions
- Deletion - Documented in this PR
- Update - No need of documentation since it is automatic.

#### Helm Container Image  (Works in Airgap mode and non-airgap mode)
- Creation - Already documented in this PR
- Deletion - Not Documented 
- Updation -  Documented in this PR.

**Review by rohitsakala — CHANGES_REQUESTED** [review]
Commented

File: docs/integrations-in-rancher/rancher-extensions.md (hunk 6 above)
Comment (martyav) [inline docs/integrations-in-rancher/rancher-extensions.md:102]:
```suggestion
### Updating and Upgrading an Extensions Repository in an Air-gapped Environment
```

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 8]
```diff
@@ -41,17 +41,41 @@ In v2.7.0, the built-in extensions will not be displayed under the **Available**
...
 
     ![Install Kubewarden](/img/install-kubewarden.png)
 
 6. Click the **Reload** page button that will appear after your extension successfully installs. Note that a logged-in user who has just installed an extension will not see a change to the UI **unless** they reload the page.
 
     ![Reload button](/img/reload-button.png)
 
+### Importing and Installing Extensions in an Air-Gapped Environment
+
+1. Find the address of the container image you want to import as an extension. Rancher provides some extensions, such as Kubewarden and Elemental, through the `ui-plugin-catalog` container image at https://hub.docker.com/r/rancher/ui-plugin-catalog/tags. You should import and use the latest tagged version of the image to ensure you receive the latest features and security updates.
```
Comment (jordojordo) [inline docs/integrations-in-rancher/rancher-extensions.md:54]: Find the address of the container image you want to import as an extension _repository_.

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 9]
```diff
@@ -41,17 +41,41 @@ In v2.7.0, the built-in extensions will not be displayed under the **Available**
...
+
+    * **(Optional)** If the container image is private: [Create](../how-to-guides/new-user-guides/kubernetes-resources-setup/secrets.md) a registry secret within the `cattle-UI-plugin-system` namespace. Enter the domain of the image address in the **Registry Domain Name** field.
```
Comment (jordojordo) [inline docs/integrations-in-rancher/rancher-extensions.md:56]: The namespace **must** be all lower case, i.e. `cattle-ui-plugin-system`.

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 10]
```diff
@@ -67,17 +91,43 @@ You must reload the page after disabling extensions or display issues may occur.
...
 :::
 
-## Rolling Back Extensions
+## Deleting Helm Charts
 
-Under the **Installed** tab, click the **Rollback** button on the extension you wish to roll back.
+1. Click **☰**, then click on the name of your local cluster.
+1. From the sidebar, select **Apps > Installed Apps**.
+1. Find the name of the chart you want to delete and select the checkbox next to it. 
+1. Click **Delete**.
```
Comment (jordojordo) [inline docs/integrations-in-rancher/rancher-extensions.md:99]: Why is this section here? You shouldn't need to delete an extension chart from the Installed Apps page.

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 11]
```diff
@@ -67,17 +91,43 @@ You must reload the page after disabling extensions or display issues may occur.
...
+1. Select the **Updates** tab. 
+1. Click **Update**.
 
-You must reload the page after rolling back extensions or display issues may occur.
+If there is a new version of the extension, there will also be an **Update** button visible on the associated card for the extension in the **Available** tab.
 
-:::
+### Updating and Upgrading an Extensions Repository in an Air-gapped Environment
+
+Extensions repository that aren't air-gapped are automatically updated. If the repository is air-gapped, you must update it manually.
```
Comment (jordojordo) [inline docs/integrations-in-rancher/rancher-extensions.md:111]: _Extension repositories_ that aren't air-gapped...

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 12]
```diff
@@ -67,17 +91,43 @@ You must reload the page after disabling extensions or display issues may occur.
...
+
+First, mirror the latest changes to your private repository by following the same steps for initially [importing and installing an extension repository](#importing-and-installing-extensions-in-an-air-gapped-environment).
```
Comment (jordojordo) [inline docs/integrations-in-rancher/rancher-extensions.md:113]: ...changes to your private _registry_ by following...

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 13]
```diff
@@ -67,17 +91,43 @@ You must reload the page after disabling extensions or display issues may occur.
...
+
+After you mirror the latest changes, follow these steps:
+
+1. Click **☰ > Local**.
+1. From the sidebar, select **Workloads > Deployments**.
+1. From the namespaces dropdown menu, select **All Namespaces**.
```
Comment (jordojordo) [inline docs/integrations-in-rancher/rancher-extensions.md:119]: It is unnecessary to select All Namespaces, you should just select the `cattle-ui-plugin-system` namespace.

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 14]
```diff
@@ -67,17 +91,43 @@ You must reload the page after disabling extensions or display issues may occur.
...
+1. Find the **cattle-ui-plugin-system** namespace. 
+1. Select the `ui-plugin-catalog` deployment.
+1. Click **⋮ > Edit config**.
+1. Update the **Container image** field with the container of the deployment.
```
Comment (jordojordo) [inline docs/integrations-in-rancher/rancher-extensions.md:123]: This is slightly confusing, it should read: "Update the Container Image field within the deployment's container with the latest image."

File: docs/integrations-in-rancher/rancher-extensions.md [hunk 15]
```diff
@@ -67,17 +91,43 @@ You must reload the page after disabling extensions or display issues may occur.
...
+1. Click **Save**.
+
+## Deleting Extension Repositories
+
+1. Click **☰ > Extensions** under **Configuration**.
+1. On the top right, click **⋮ > Manage Repositories**.
+1. Find the name of the extension repository you want to delete. Select the checkbox next to the repository name, then click **Delete**.
```
Comment (jordojordo) [inline docs/integrations-in-rancher/rancher-extensions.md:130]: This section should provide instructions on deleting the Extension Catalog, not the repository. Deleting an imported extension catalog will automatically remove the associated repository, deployment, and service.

**Review by jordojordo — CHANGES_REQUESTED** [review]

File: docs/integrations-in-rancher/rancher-extensions.md (hunk 15 above)
Comment (jordojordo) [inline docs/integrations-in-rancher/rancher-extensions.md:130]: Is this related to Air-gapped extension repositories that are created when importing an extension catalog?

**Review by jordojordo — CHANGES_REQUESTED** [review]

File: docs/integrations-in-rancher/rancher-extensions.md (hunk 8 above)
Comment (martyav) [inline docs/integrations-in-rancher/rancher-extensions.md:54]:
```suggestion
1. Find the address of the container image repository  that you want to import as an extension. Rancher provides some extensions, such as Kubewarden and Elemental, through the `ui-plugin-catalog` container image at https://hub.docker.com/r/rancher/ui-plugin-catalog/tags. You should import and use the latest tagged version of the image to ensure you receive the latest features and security updates.
```

File: docs/integrations-in-rancher/rancher-extensions.md (hunk 9 above)
Comment (martyav) [inline docs/integrations-in-rancher/rancher-extensions.md:56]:
```suggestion
    * **(Optional)** If the container image is private: [Create](../how-to-guides/new-user-guides/kubernetes-resources-setup/secrets.md) a registry secret within the `cattle-ui-plugin-system` namespace. Enter the domain of the image address in the **Registry Domain Name** field.
```

File: docs/integrations-in-rancher/rancher-extensions.md (hunk 11 above)
Comment (martyav) [inline docs/integrations-in-rancher/rancher-extensions.md:111]:
```suggestion
Extensions repositories that aren't air-gapped are automatically updated. If the repository is air-gapped, you must update it manually.
```

File: docs/integrations-in-rancher/rancher-extensions.md (hunk 12 above)
Comment (martyav) [inline docs/integrations-in-rancher/rancher-extensions.md:113]:
```suggestion
First, mirror the latest changes to your private registry by following the same steps for initially [importing and installing an extension repository](#importing-and-installing-extensions-in-an-air-gapped-environment).
```

File: docs/integrations-in-rancher/rancher-extensions.md (hunk 13 above)
Comment (martyav) [inline docs/integrations-in-rancher/rancher-extensions.md:119]:
```suggestion
1. From the namespaces dropdown menu, select **cattle-ui-plugin-system**.
```

File: docs/integrations-in-rancher/rancher-extensions.md (hunk 14 above)
Comment (martyav) [inline docs/integrations-in-rancher/rancher-extensions.md:123]:
```suggestion
1. Update the **Container Image** field within the deployment's container with the latest image.
```

**Comment by rohitsakala** [comment]
**CRUD of Extensions Repository**

#### Helm Github Repository (Works in non-airgap mode only)
- Creation - Already documented in https://ranchermanager.docs.rancher.com/v2.8/integrations-in-rancher/rancher-extensions
- Deletion - Documented in this PR
- Update - No need for documentation since it is automatic.

#### Helm Container Image  (Works in Airgap mode and non-airgap mode)
- Creation - Already documented in this PR
- Deletion - Not Documented 
- Updation -  Documented in this PR.

@martyav The deletion of Extension Repository through Container Image (helm) is not yet documented. Rest of LGTM.

**Review by rohitsakala — CHANGES_REQUESTED** [review]
Commented

**Comment by rohitsakala** [comment]
@martyav There is an option to `Uninstall` the Extension repo container image. 

Screenshot 2023-11-28 at 1 15 19 PM

**Pull request merged**